packages/jest-worker/src/index.js

```diff
@@ -0,0 +1,193 @@
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+'use strict';
+
+const mergeStream = require('merge-stream');
+const os = require('os');
+const path = require('path');
+
+const {CHILD_MESSAGE_CALL, CHILD_MESSAGE_END} = require('./types');
+const Worker = require('./worker');
+
+import type {FarmOptions} from './types';
+import type {Readable} from 'stream';
+
+/**
+ * The Jest farm (publicly called "Worker") is a class that allows you to queue
+ * methods across multiple child processes, in order to parallelize work. This
+ * is done by providing an absolute path to a module that will be loaded on each
+ * of the child processes, and bridged to the main process.
+ *
+ * Bridged methods are specified by using the "exposedMethods" property of the
+ * options "object". This is an array of strings, where each of them corresponds
+ * to the exported name in the loaded module.
+ *
+ * You can also control the amount of workers by using the "numWorkers" property
+ * of the "options" object, and the settings passed to fork the process through
+ * the "forkOptions" property. The amount of workers defaults to the amount of
+ * CPUS minus one.
+ *
+ * Queueing calls can be done in two ways:
+ *   - Standard method: calls will be redirected to the first available worker,
+ *     so they will get executed as soon as they can.
+ *
+ *   - Sticky method: if a "computeWorkerKey" method is provided within the
+ *   config, the resulting string of this method will be used as a key.
+ *   Everytime this key is returned, it is guaranteed that your job will be
+ *   processed by the same worker. This is specially useful if your workers are
+ *   caching results.
+ */
+class Farm {
+  _stdout: Readable;
+  _stderr: Readable;
+  _ending: boolean;
+  _cacheKeys: {[string]: Worker};
+  _options: FarmOptions;
+  _workers: Array<Worker>;
+
+  constructor(workerPath: string, options?: FarmOptions = {}) {
+    const numWorkers = options.numWorkers || os.cpus().length - 1;
+    const workers = new Array(numWorkers);
+    const stdout = mergeStream();
+    const stderr = mergeStream();
+
+    if (!path.isAbsolute(workerPath)) {
+      workerPath = require.resolve(workerPath);
+    }
+
+    // Build the options once for all workers to avoid allocating extra objects.
+    const workerOptions = {
+      forkOptions: options.forkOptions || {},
+      workerPath,
+    };
+
+    for (let i = 0; i < numWorkers; i++) {
+      const worker = new Worker(workerOptions);
+
+      stdout.add(worker.getStdout());
+      stderr.add(worker.getStderr());
+
+      workers[i] = worker;
+    }
+
+    let exposedMethods = options.exposedMethods;
+
+    // If no methods list is given, try getting it by auto-requiring the module.
+    if (!exposedMethods) {
+      // $FlowFixMe: This has to be a dynamic require.
+      const child = require(workerPath);
+
+      exposedMethods = Object.keys(child)
+        .filter(name => this._isValidMethod(name))
```

Comment:
This makes it silently ignore `end`. Should we throw if `end` is provided?

Reply:
I thought about that; but what would happen if you require a module who exposes an `end` method, or any other underscored property we would crash. If you have a strong opinion on that I can remove the filter for validity.

Reply:
I think we should throw, and in those cases demand people to use `exposedMethods` or wrap their module. Otherwise it's super confusing.

Reply:
Oki!